### PR TITLE
Fork process for background mounting

### PR DESCRIPTION
**Pull request by mtotg**
What does this PR change? What problem does it solve?
-----------------------------------------------------

This allows restic to be forked and put into the background for mounts to continue running even when restic closes.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
Issue #2743 discusses this.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by mtotg** [comment]
This is a WIP, but it's an initial pass. Though, I realize this is a recursive loop (code below), need to strip the DaemonMode argument off the args array.

```
pid, err := syscall.ForkExec(filename, args, attr)
```

File: cmd/restic/cmd_mount.go [hunk 1]
```diff
@@ -79,9 +81,38 @@ func init() {
...
+	if opts.DaemonMode {
+		filename := os.Args[0]
+		attr := new(syscall.ProcAttr)
+		argx := removeElement(args, "daemon")
+		pid, err := syscall.ForkExec(filename, argx, attr)
+		Println("Mount is now running as a daemon. PID: ", pid)
+		if err != nil {
+			panic(err)
+		}
+		return errors.New("Daemon mode running in background")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_mount.go:114]: The exit status should be zero if the background mount succeeded.

File: cmd/restic/cmd_mount.go [hunk 2]
```diff
@@ -79,9 +81,38 @@ func init() {
...
+func linearSearch(args []string, search string) int {
+	for i, n := range args {
+		if n == search {
+			return i
+		}
+	}
+	return -1
+}
+
+func removeElement(args []string, search string) []string {
```
Comment (MichaelEischer) [inline cmd/restic/cmd_mount.go:96]: The function interface is misleading as `args` is modified in place.

File: cmd/restic/cmd_mount.go [hunk 3]
```diff
@@ -79,9 +81,38 @@ func init() {
...
+		return append(args[:index], args[index+1:]...)
+	}
+	return args
 }
 
 func runMount(opts MountOptions, gopts GlobalOptions, args []string) error {
+	if opts.DaemonMode {
+		filename := os.Args[0]
+		attr := new(syscall.ProcAttr)
+		argx := removeElement(args, "daemon")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_mount.go:108]: The `args` parameter has already been parsed by cobra. Calls like `restic -r repo mount --daemon` probably don't work?

File: cmd/restic/cmd_mount.go [hunk 4]
```diff
@@ -164,12 +195,13 @@ func runMount(opts MountOptions, gopts GlobalOptions, args []string) error {
 	Printf("When finished, quit with Ctrl-c or umount the mountpoint.\n")
 
 	debug.Log("serving mount at %v", mountpoint)
+
```
Comment (MichaelEischer) [inline cmd/restic/cmd_mount.go:198]: Please avoid unnecessary whitespace changes.

**Comment by greatroar** [comment]
The reporter of #2743 wants to "*immediately* use the mounted repo once it is successfully mounted." For that to work, the parent process needs to wait for the child to report a successful mount, probably through a pipe. Do you intend to implement that?

File: cmd/restic/cmd_mount.go (hunk 3 above)
Comment (greatroar) [reply cmd/restic/cmd_mount.go:108]: Instead of removing an argument, maybe the child process should have a hidden flag `--daemon-subprocess`. I'm not sure if cobra support hidden flags, though. It could take a key shared with the parent as an argument, to ensure it doesn't work from the command line. `time.Now().UnixNano()` in hex should do.

File: cmd/restic/cmd_mount.go [hunk 5]
```diff
@@ -79,9 +81,38 @@ func init() {
...
+	if index != -1 {
+		return append(args[:index], args[index+1:]...)
+	}
+	return args
 }
 
 func runMount(opts MountOptions, gopts GlobalOptions, args []string) error {
+	if opts.DaemonMode {
+		filename := os.Args[0]
+		attr := new(syscall.ProcAttr)
```
Comment (greatroar) [inline cmd/restic/cmd_mount.go:107]: A daemon should start a new session and chdir to the root:

```go
attr := &syscall.ProcAttr{
	Dir: "/",
	Sys: &syscall.SysProcAttr{Setsid: true},
}
```

Maybe Noccty should be set as well?

**Comment by mtotg** [comment]
> The reporter of #2743 wants to "_immediately_ use the mounted repo once it is successfully mounted." For that to work, the parent process needs to wait for the child to report a successful mount, probably through a pipe. Do you intend to implement that?

Given that it’s forked and then needs to do the mounting cycle, my intent wasn’t to implement a pipe back to the main process to signal that though. Additionally, I’m not sure a pipe would be the best way if implemented.

**Comment by greatroar** [comment]
Without signaling back, isn't this PR just a complicated way to do `restic mount &`?

**Comment by MichaelEischer** [comment]
Closing this PR as it just seems to implement `restic mount &`.
